src/components/EnhancedMap/ImageMarkerLayer/ImageMarkerLayer.jsx

```diff
@@ -50,85 +64,129 @@ const ImageMarkerLayer = props => {
   images: PropTypes.arrayOf(PropTypes.shape({
-    key: PropTypes.string,
-    url: PropTypes.string,
-    position: PropTypes.object,
-  })),
-  imageClicked: PropTypes.func,
+    key: PropTypes.string.isRequired,
+    url: PropTypes.string.isRequired,
+    position: PropTypes.shape({
+      lat: PropTypes.number.isRequired,
+      lon: PropTypes.number.isRequired,
+    }).isRequired,
+  })).isRequired,
+  imageClicked: PropTypes.func.isRequired,
   imageAlt: PropTypes.string,
   buildIcon: PropTypes.func,
   markerColor: PropTypes.string,
 }
 
-class MapillaryViewer extends Component {
-  containerRef = createRef();
+const MapillaryViewer = ({ initialImageKey }) => {
+  const containerRef = useRef()
 
-  componentDidMount() {
-    this.viewer = new Viewer({
-      accessToken: getAccessToken(),
-      container: this.containerRef.current,
-      imageId: this.props.initialImageKey,
-      component: { cover: false },
-    })
-  }
+  useEffect(() => {
+    if (!initialImageKey) {
+      console.error('Initial image key is null or undefined')
+      return
+    }
+
+    let viewer
+    try {
+      const accessToken = getAccessToken();
+      if (!accessToken) {
+        console.error('Access token is not available');
+        return;
+      }
 
-  componentWillUnmount() {
-    if (this.viewer) {
-      this.viewer.remove();
+      if (imageCache.has(initialImageKey)) {
+        viewer = imageCache.get(initialImageKey);
+      } else {
+        viewer = new Viewer({
+          accessToken: accessToken,
+          container: containerRef.current,
+          imageId: initialImageKey,
+        });
+        imageCache.set(initialImageKey, viewer);
+      }
+    } catch (error) {
+      console.error('Error initializing Mapillary viewer:', error)
     }
-  }
 
-  shouldComponentUpdate(nextProps) {
-    return nextProps.initialImageKey !== this.props.initialImageKey
-  }
+    return () => {
+      try {
+        if (viewer) {
+          viewer.remove()
+        }
+      } catch (error) {
+        console.error('Error removing Mapillary viewer:', error)
+      }
+    }
+  }, [initialImageKey])
 
-  render() {
-    return (
-      <div className="mr-p-2 mr-pt-4 mr-relative">
-        <div ref={this.containerRef} id="mapillary-viewer" style={{ width: 335, height: 263 }}></div>
-      </div>
-    )
-  }
+  return (
+    <div className="mr-p-2 mr-pt-4 mr-relative">
+      <div ref={containerRef} id="mapillary-viewer" className="mr-w-full mr-h-64"></div>
+    </div>
+  )
 }
 
-const buildImageMarkers = (images, icon, imageClicked, imageAlt, layerId, layerLabel) => {
-  if (!images || images.length === 0) {
+const buildImageMarkers = (images, icon, markerColor, imageClicked, layerId, layerLabel) => {
+  try {
+    if (!images || images.length === 0) {
+      return []
+    }
+
+    return _map(images, imageInfo => {
+      if (!imageInfo || (!imageInfo.position?.lat || !imageInfo.position?.lon) && (!imageInfo.lat || !imageInfo.lon)) {
```

Comment:
This check has a minor bug, because you're coercing a number (`imageInfo.position?.lat`) to a boolean, which doesn't do what you want if the number == 0. In most cases you want to explicitly compare the value with `undefined` instead (e.g. `imageInfo.position?.lat !== undefined)

It's also pretty hard to read this condition because of its complexity; I'd suggest maybe extracting some of the logic into a `getLatLng(imageInfo) -> [number, number]?` function and then making your condition `if (!imageInfo || !getLatLng(imageInfo))`.